kernel/inferCumulativity.ml

```diff
@@ -227,12 +227,12 @@ let rec infer_fterm cv_pb infos variances hd stk =
       infer_inductive_instance cv_pb (info_env (fst infos)) variances ind nargs u
     in
     infer_stack infos variances stk
-  | FConstruct (ctor,u) ->
+  | FConstruct ((ctor,u),args) ->
     let variances =
-      let nargs = stack_args_size stk in
+      let nargs = Array.length args in
```

Comment:
Just to err on the safe side I'd rather use `Array.length args + stack_args_size stk`. Due to zupdates we can't be sure that the application is full. There are some other places like that.

Reply:
zip can update stuff, and it runs after knh. I don't see how you can fix it there.

Reply:
I guess you mean the zip in whd_stack, but if kni returned a fconstruct then the stack is empty so zip did nothing.

Reply:
OK, maybe it's fine for zip, but there are other places where we update Zupdate nodes. I don't see why we would magically maintain the invariant that constructor nodes are fully applied.

Reply:
the invariant is that constructor nodes at the head of the result of reduction are fully applied, not constructor nodes in random positions in the stack or wherever